Add an internal refresh api that makes pipeline changes visible on all ingest nodes

martijnvg opened the pull request:
instead of only relying on async background task that periodically loads pipelines into memory. This async loading mechanism remains for loading pipelines during startup.

* Pipeline store can now only be used when there is no .ingest index or all primary shards of .ingest have been started
* IngestPlugin adds`node.ingest` setting to `true`. This is used to figure out to what nodes to send the refresh request too. This setting isn't yet configurable. This will be done in a follow up issue.
* Async background loading mechanism only runs now every 30 minutes instead of every second.

martijnvg commented:
I've updated this PR to move some initialisation logic out of the store into a new class called PipelineStoreBootstrapper. The pipeline store became a bit convoluted with initialisation logic.

talevy commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/refresh/TransportRefreshAction.java:
```diff
@@ -0,0 +1,117 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Refreshes the in-memory representaion of all the pipelines on all ingest nodes.
```
s/representaion/representa`t`ion/

talevy commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/refresh/TransportRefreshAction.java:
```diff
@@ -0,0 +1,117 @@
+        this.pipelineStore = pipelineStore;
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        transportService.registerRequestHandler(RefreshAction.NAME, RefreshRequest::new, ThreadPool.Names.SAME, this);
```
just for my own education, what is the `ThreadPool.Names.SAME` pool supposed to represent?

uboness replied:
it's a thread pool that is not really a thread pool :)... the callables/runnables will be executed in the same thread that added them to the pool. In this case, I believe that will be the network thread.

talevy replied:
interesting, so it is completely unrelated to the threadpool of the response handler (in this case `ThreadPool.Names.MANAGEMENT`).

thanks

talevy commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -60,73 +63,82 @@
-    protected void doClose() {
-        // TODO: When org.elasticsearch.node.Node can close Closable instances we should remove this code
+    public void close() throws IOException {
+        // TODO: When org.elasticsearch.node.Node can close Closable instances we should try to remove this code,
```
is there a reference issue to keep track of this to-be-introduced functionality?

martijnvg replied:
I think not specifically for this. At some point in core a lot of the guice stuff is going to disappear / be refactored. One thing that what I understood is that a node will close all services that implement Closable and LifeCycleAbstractComponent is going to get removed.

talevy commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStoreBootstrapper.java:
```diff
@@ -0,0 +1,101 @@
+        this.threadPool = threadPool;
+        this.pipelineStore = pipelineStore;
+        this.clusterService = clusterService;
+        this.pipelineUpdateInterval = settings.getAsTime("ingest.pipeline.store.update.interval", TimeValue.timeValueMinutes(30));
```
maybe we should add this to the ingest docs?

martijnvg replied:
+1 makes sense

talevy commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -33,12 +33,11 @@
 import org.elasticsearch.cluster.ClusterChangedEvent;
```
lots of unused imports can be removed here.

talevy commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -228,17 +275,28 @@ synchronized void updatePipelines() throws Exception {
+    }
+
+    private Iterable<SearchHit> readAllPipelines() {
         // TODO: the search should be replaced with an ingest API when it is available
```
should we open an issue for this?

martijnvg replied:
Yes, that makes sense. I think this api should be named list pipeline API? Which under the hood just executes a scroll search. Although I think it will be phase2?

talevy commented:
would it be beneficial to have refresh action/request tests in `IngestClientIT`?

martijnvg commented:
I think it makes sense to add a test for this api, even though this is an internal api that is designed to be used only via put and delete pipeline apis. The reason that I did not add an integration test for this yet, is that it is already tests by the put and delete pipeline rest and ClientIT tests.

talevy commented:
but these are just one-node tests, no? since this gets info from cluster-state about multiple nodes (some of which may not be ingest nodes), maybe it makes sense to test those cases as well

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -60,73 +63,82 @@
+        client().delete(deleteRequest, new ActionListener<DeleteResponse>() {
+            @Override
+            public void onResponse(DeleteResponse deleteResponse) {
+                client().execute(RefreshAction.INSTANCE, new RefreshRequest(), new ActionListener<RefreshResponse>() {
```
if the api is really internal, I think we can simplify this. Do we need to use a client here? Can we instead use the transport service directly? In that case we wouldn't need the RefreshAction, and the RefreshRequestBuilder. Otherwise the api ends up being exposed anyways, no matter if we say it's internal.

martijnvg replied:
agreed and this what I initially tried (like the shards exists internal api, See IndicesStore.java), but then I ran into cycle dependency issue with transport service.

javanna replied:
ok...but client depends on the transport service anyway no? I think I don't get it

martijnvg replied:
yes it does. all the ingest transport stuff gets instantiate with the the
action module and not with the node modules of the plugin. (See
IngestPlugin#onModule(ActionModule) ) So that is why the current approach
works.

On 4 December 2015 at 12:47, Luca Cavanna <notifications@github.com> wrote:

> In
> plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java
> <https://github.com/elastic/elasticsearch/pull/15203#discussion_r46674077>
> :
>
> >          DeleteRequest deleteRequest = new DeleteRequest(request);
> >          deleteRequest.index(PipelineStore.INDEX);
> >          deleteRequest.type(PipelineStore.TYPE);
> >          deleteRequest.id(request.id());
> >          deleteRequest.refresh(true);
> > -        client().delete(deleteRequest, listener);
> > +        client().delete(deleteRequest, new ActionListener<DeleteResponse>() {
> > +            @Override
> > +            public void onResponse(DeleteResponse deleteResponse) {
> > +                client().execute(RefreshAction.INSTANCE, new RefreshRequest(), new ActionListener<RefreshResponse>() {
>
> ok...but client depends on the transport service anyway no? I think I
> don't get it
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/15203/files#r46674077>.
>



-- 
Met vriendelijke groet,

Martijn van Groningen

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -60,73 +63,82 @@
+                client().execute(RefreshAction.INSTANCE, new RefreshRequest(), new ActionListener<RefreshResponse>() {
+                    @Override
+                    public void onResponse(RefreshResponse refreshResponse) {
+                        // TODO: do we want to include in the response how many ingest nodes successfully refreshed their pipelines?
```
I don't think it's important for now

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -60,73 +63,82 @@
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        logger.error("Failed to update pipelines on other ingest nodes [{}]", e);
```
did you plan to add here the list of nodes or something? looks like there is a missing argument.

martijnvg replied:
oops, I'll add this.

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -60,73 +63,82 @@
+                    public void onFailure(Throwable e) {
+                        logger.error("Failed to update pipelines on other ingest nodes [{}]", e);
+                        // TODO: do we want to fail the request here?
+                        // We did manage to successfully store the pipeline document, but ingest nodes where unable to load
```
s/where/were  in the comment.

I am not sure what we should do here. Change is stored but in order for it to be seen we need to wait for the next refresh in the background. Failing might sound like a good idea, but then people can't do anything about it, cause the refresh api is internal and can't be called through REST.

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -228,17 +275,28 @@ synchronized void updatePipelines() throws Exception {
-        threadPool.schedule(pipelineUpdateInterval, ThreadPool.Names.GENERIC, new Updater());
-    }
-
     boolean existPipeline(String pipelineId) {
```
doesn't need to be part of this PR necessarily but maybe we should rename to pipelineExists ?

martijnvg replied:
+1 that sounds better

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/refresh/TransportRefreshAction.java:
```diff
@@ -0,0 +1,117 @@
+        } catch (Exception e) {
+            logger.error("unable to update pipelines on node [{}]", e, clusterService.localNode());
+        }
+        channel.sendResponse(new RefreshResponse(true));
```
if things fail, don't we want to return false?

martijnvg replied:
yes, we do, i'll change this

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/refresh/TransportRefreshAction.java:
```diff
@@ -0,0 +1,117 @@
+
+                @Override
+                public void handleResponse(RefreshResponse response) {
+                    decrementAndReturn();
```
we ignore whether refresh failed or not. (cause it's at the moment always true) Wondering if we should take it into account.

javanna commented on plugins/ingest/src/test/resources/rest-api-spec/test/ingest/30_grok.yaml:
```diff
@@ -21,14 +21,6 @@
-        wait_for_nodes: 99
-        timeout: 2s
-  - match: { "timed_out": true }
-
```
seeing all these sleeps going away from REST tests is a huge relief!

s1monw commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -60,73 +63,82 @@
 
     @Inject
-    public PipelineStore(Settings settings, Provider<Client> clientProvider, ThreadPool threadPool, Environment environment, ClusterService clusterService, Map<String, ProcessorFactoryProvider> processorFactoryProviders) {
+    public PipelineStore(Settings settings, Provider<Client> clientProvider, Environment environment, Map<String, ProcessorFactoryProvider> processorFactoryProviders) {
```
what is this provider doing here? We can't do this in our production code this will take hours for decouple again We either pass a client or not but not yet another indirection.

s1monw commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -149,13 +163,39 @@ public void put(PutPipelineRequest request, ActionListener<IndexResponse> listen
         indexRequest.source(request.source());
         indexRequest.refresh(true);
-        client().index(indexRequest, listener);
+        client().index(indexRequest, new ActionListener<IndexResponse>() {
```
all these listeners look the same, can we streamline it?

s1monw commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStoreBootstrapper.java:
```diff
@@ -0,0 +1,101 @@
+        }
+    }
+
+    void scheduleBackgroundUpdater() {
```
why do we still have this? isn't the purpose of this change to remove this?

martijnvg replied:
The refresh (which has now been renamed to reload pipelines) works well for nodes that are currently running. But for a node that is currently starting up and in process loading the pipelines it might just miss a pipeline update. This will then be picked up the next time the background updater runs.

s1monw replied:
can't we fix this differently, I mean once we joined the cluster all nodes should have the new clusterstate so we are visible. That means if we join while a node is sending out reloads based on a given clusterstate we can just check after we are doing if the clusterstate has changed and re-reload if a new ingest node joined?

martijnvg replied:
Let me think... when a node loads the pipelines via scroll search it is already part of the cluster. What I'm concerned about is visibility of pipelines changes in the `.ingest` index. So a node is scroll searching for pipelines, an update is made to a pipeline on some other node. On the node that is loading pipelines both of these actions go through the `PipelineStore#updatePipelines()` method. This method is synchronized, so in this case the reload operation caused by a pipeline update will have to wait until the initial pipeline loading has completed. So my conclusion is that this is safe and we can remove the scheduling of background update tasks.

Does my reasoning make sense?

martijnvg replied:
Beyond the scenario described above there other scenarios we need to consider:
* An ingest node gets isolated from the cluster. When it gets back, it needs to check for pipeline changes.
* The scroll search that fetches the pipeline fails. We need to retry. 

We can add specific logic for both cases, but the current background updater does make sure that eventually the pipeline changes are pushed to all ingest nodes.

kimchy replied:
We can come up with a change to make sure we remove the scheduled reload, but it is not simple. My suggestion, this change is already a great step forward in terms of testability and visibility, and we can keep this simple in this change while relying on the background change. In a different change, we can think about an algo to remove the need for scheduled refresh?

s1monw replied:
maybe we can keep this very simple and if we can't find a pipeline we just try to fetch that missing pipeline? that would simplify all the stuff here dramatically no?

martijnvg replied:
How would we then deal with updates made to existing pipelines? Fetch pipelines all the time?

s1monw replied:
all I want here is a mechanism that always works. I think if we rely on a backgroud task our system is broken and we have to fix it.

javanna replied:
> all I want here is a mechanism that always works.

we have this common goal, I agree that not having a background task would be cleaner, but I wonder how we can address all the different problems of a distributed system (a node loses connectivity and comes back up, the search for pipelines returns an error e.g. shard not available etc.) with a sync approach. Sounds like moving completely to sync at the moment means that we will potentially not find pipelines or execute not up-to-date pipelines... Maybe we should identify changes that have to be made to core so that everything can be made sync, but in the meantime, this PR is already a big improvement as it makes things testable and at least partially sync.

martijnvg replied:
@s1monw @javanna @kimchy I can improve the pipeline update logic, so that we handle the cases that I've described (ingest node isolation from the cluster, search failures) correctly. Then I think the background update thread can be removed. (unless I missed other cases) So would this be the right step forward?

s1monw commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -214,7 +225,7 @@ synchronized void updatePipelines() throws Exception {
         int removed = 0;
         for (String existingPipelineId : pipelines.keySet()) {
-            if (!existPipeline(existingPipelineId)) {
+            if (!pipelineExists(existingPipelineId)) {
```
can we use `== false`

martijnvg replied:
ok, will change.

martijnvg commented:
@javanna @talevy I've updated the PR. Changed:
* Renamed refresh action to reload pipelines action.
* Made the reload pipelines action an internal api (Java clients cannot invoke it).
* Removed the `refreshed` field from the response. Whether the reload was successful is ignored and only logged. We should remain to rely on the background pipeline reloading mechanism.
* Set the reload pipeline background task interval from 30 minutes to 30 seconds.

s1monw commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/reload/ReloadPipelinesAction.java:
```diff
@@ -0,0 +1,137 @@
+    private PipelineStore pipelineStore;
+
+    @Inject
+    public ReloadPipelinesAction(Settings settings, Provider<PipelineStore> pipelineStoreProvider, ClusterService clusterService, TransportService transportService) {
```
sorry but we have to find other solutions for this than using these injection providers. We can't sustain this kind of software engineering here.

martijnvg replied:
This one is easy. Can make this an inner class of PipelineStore and then there isn't a cyclic dependency.

s1monw replied:
why is shit guice exposed at all? Can't we reduce the number of @Inject here please it's such a pain to unwire all of this. Can we simply remove the @Inject and all the wiring and to in `PipelineStoreBootstrapper`

```Java

ReloadPipelinesAction action = new ReloadPipelinesAction(settings, pipelineStore, clusterService, transportService);

```

martijnvg replied:
so the reason pipeline store instance is managed by Guice is that the transport action classes uses it as a dependency. But that shouldn't be a problem, because if the store is exposed as a getter and the transport action classes depend on `PipelineStoreBootstrapper` then the constructor of the transport action classes can use this getter.

s1monw commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/reload/ReloadPipelinesAction.java:
```diff
@@ -0,0 +1,137 @@
+
+    private PipelineStore pipelineStore() {
+        if (pipelineStore == null) {
+            pipelineStore = pipelineStoreProvider.get();
```
this method is entirely unused and you don't need that `pipelineStoreProvider` at all

s1monw replied:
oh nevermind

martijnvg commented:
@s1monw I've updated this PR. Only the `PipelineStoreBootstrapper` is now managed by Guice. I'm still using a Guice `Provider` for the `Client` dependency. There is a cyclic dependency, the bootstrapper class is injected in the transport action classes, but Client is also injected into the bootstrapper class. I'm not sure how to get around this. 

This is a general problem. If a component is used in a transport action in a plugin and that component either directly or indirectly depends on Client then there is a cyclic dependency. They way we have been dealing with this by injecting client lazily (either by using a Guice `Provider`or directly using Guice's injector).

s1monw commented:
I will look and try to help @martijnvg

s1monw commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -228,17 +217,32 @@ synchronized void updatePipelines() throws Exception {
+        try {
+            updatePipelines();
+        } catch (Exception e) {
+            logger.error("Failed to load the pipelines during startup");
```
please log the exception here as well we really wanna see what was going wrong

martijnvg replied:
oops I missed that, I've added the exception.

martijnvg commented:
I've updated this PR and removed the background pipeline updater and added logic to deal with master blocks and when the pipeline loading itself fails.

talevy commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/reload/ReloadPipelinesAction.java:
```diff
@@ -0,0 +1,123 @@
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.warn("failed to update pipelines on remote node [{}]", exp, node);
```
It would be great to add a test for this scenario.

An ingest node may have disconnected or randomly stopped being an ingest node between the time that 
`node.getAttributes().get("ingest")` was called, and when the `ReloadPipelineRequest` was actually issued against it.

javanna commented on plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java:
```diff
@@ -200,9 +184,15 @@ synchronized void updatePipelines() throws Exception {
+            PipelineDefinition current = newPipelines.get(pipelineId);
+            if (current != null) {
+                // If we first read from a primary shard copy and then from a replica copy,
+                // and a write did not yet make it into th replica shard
```
s/th/the

javanna commented on docs/plugins/ingest.asciidoc:
```diff
@@ -474,8 +474,10 @@ PUT _ingest/pipeline/my-pipeline-id
+NOTE: The put pipeline api also instructs all ingest nodes to reload their in-memory representation of pipelines, so that
+      pipeline changes take in effect immediately. Also each ingest node asynchronously checks for pipeline changes in the
+      background. How often processors are updated in the background can be controlled via the
+      `ingest.pipeline.store.update.interval` setting and defaults is `30s`.
```
this needs to be updated if we go for removing the background updater completely.

martijnvg commented:
@javanna @s1monw I've updated this PR. I also replaced the Injector<Client> workaround by a client setter instead. This is cleaner, but in an ideal world Client should be a constructor parameter, but that requires some changes how setup `NodeClient`.

martijnvg commented:
I opened #15504 to track the cyclic dependency issues.

javanna commented:
LGTM
